### PR TITLE
[SPARK-16030] [SQL] Allow specifying static partitions when inserting to data source tables

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

This PR adds the static partition support to INSERT statement when the target table is a data source table.
## How was this patch tested?

New tests in `InsertSuite`

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 1]
```diff
@@ -284,4 +285,78 @@ class InsertSuite extends DataSourceTest with SharedSQLContext {
...
+      checkAnswer(
+        sql("SELECT a, b, c, d FROM srcSparkPart"),
+        Row(1, 2, 3, 4) ::
+        Row(5, 6, 7, 8) ::
+        Row(9, 10, 11, 12) ::
+        Row(13, 14, 15, 16) ::
+        Row(17, 18, 19, 20) ::
+        Row(21, 22, 23, 24) ::
+        Row(25, 26, 27, 28) :: Nil
+      )
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala]: We also need this test for hive format tables. I will add that in a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -320,6 +320,19 @@ trait CheckAnalysis extends PredicateHelper {
...
 
+          case i @ InsertIntoTable(table, partitions, query, _, _) =>
+            // TODO: We need to consolidate this kind of pre-write checks with the rule of
+            // PreWriteCheck defined in extendedCheckRules.
+            val numStaticPartitions = partitions.values.count(_.isDefined)
+            if (table.output.size != (query.output.size + numStaticPartitions)) {
+              failAnalysis(
+                s"$table requires that the data to be inserted have the same number of " +
+                s"columns as the target table: target table has ${table.output.size} " +
+                s"column(s) but the inserted data has " + "" +
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Why the `""` here?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: accident :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 3]
```diff
@@ -43,11 +44,94 @@ import org.apache.spark.unsafe.types.UTF8String
...
+      caseInsensitiveResolution
+    }
+  }
+
   override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case i @ logical.InsertIntoTable(
-           l @ LogicalRelation(t: HadoopFsRelation, _, _), part, query, overwrite, false)
-        if query.resolved && t.schema.asNullable == query.schema.asNullable =>
+    l @ LogicalRelation(t: HadoopFsRelation, _, _), parts, query, overwrite, false)
+      if query.resolved && parts.exists(_._2.isDefined) =>
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Need doc

**Comment by rxin** [comment]
This is ok for 2.0, but for something like this we should write it in a way so the rewriting logic can be unit tested without end-to-end suites.

**Comment by rxin** [comment]
Actually can you try do it in this? I worry similar to past pull requests we will never revisit and the code will just look like this 2 years from now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 4]
```diff
@@ -43,11 +44,116 @@ import org.apache.spark.unsafe.types.UTF8String
...
+    //
+    // Basically, we will put those partition columns having a assigned value back
+    // to the SELECT clause. The output of the SELECT clause is organized as
+    // normal_columns static_partitioning_columns dynamic_partitioning_columns.
+    // static_partitioning_columns are partitioning columns having assigned
+    // values in the PARTITION clause (e.g. b in the above example).
+    // dynamic_partitioning_columns are partitioning columns that do not assigned
+    // values in the PARTITION clause (e.g. c in the above example).
+    case i @ logical.InsertIntoTable(
+    l @ LogicalRelation(t: HadoopFsRelation, _, _), parts, query, overwrite, false)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: l is a bad name to use

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: and don't use t or i here either ...

**Comment by yhuai** [comment]
Closing this. https://github.com/apache/spark/pull/13769 is the new version.

**Pull request closed without merging**